apps/sms/views/shared/js/system_message_handler.js

```diff
@@ -0,0 +1,341 @@
+      return navigator.mozApps.getSelf().then((app) => {
+        app.launch();
+
+        // If we're currently in the target conversation, just do nothing.
+        if (Navigation.isCurrentPanel('thread', { id: threadId })) {
+          return;
+        }
+
+        //Validate if message still exists before opening message thread
+        //See issue https://bugzilla.mozilla.org/show_bug.cgi?id=837029
+        return Utils.onceDocumentIsVisible().then(
+          () => MessageManager.getMessage(id)
+        ).then(
+          (message) => Navigation.toPanel('thread', { id: message.threadId }),
+          () => Utils.alert('deleted-sms')
+        );
+      });
+    },
+
+    /**
+     * Handles "sms-received" system message.
+     * @param {MozMmsMessage|MozSmsMessage} message Received message.
+     * @returns {Promise.<void>} Promise that is resolved once message is fully
+     * handled.
+     * @private
+     */
+    onSmsReceivedSystemMessage(message) {
+      // Acquire the cpu wake lock when we receive an SMS.  This raises the
+      // priority of this process above vanilla background apps, making it less
+      // likely to be killed on OOM.  It also prevents the device from going to
+      // sleep before the user is notified of the new message.
+      //
+      // We'll release it once we display a notification to the user.  We also
+      // release the lock after 30s, in case we never run the notification code
+      // for some reason.
+      this.requestWakeLock();
+
+      var handleReceivedMessagePromise;
+      if (message.messageClass === 'class-0') {
+        handleReceivedMessagePromise = this.onFlashMessageReceived(message);
+      } else {
+        handleReceivedMessagePromise = SilentSms.checkSilentModeFor(
+          message.sender
+        ).then((isSilent) => {
+          if (!isSilent) {
+            return this.onMessageReceived(message);
+          }
+        });
+      }
+
+      return handleReceivedMessagePromise.then(() => this.releaseWakeLock());
+    },
+
+    /**
+     * Handles "notification" system message.
+     * @param {Notification} notification Notification object.
+     * @returns {Promise.<void>} Promise that is resolved once notification
+     * system message is handled.
+     * @private
+     */
+    onNotificationSystemMessage(notification) {
+      // When notification is removed from notification tray, notification
+      // system message will still be fired, but "clicked" property will be
+      // equal to false. This should change once bug 1139363 is landed. When
+      // user clicks on notification we'll get two system messages, first to
+      // notify app that notification is clicked and then, once we show
+      // Conversation view to the user, we remove that notification from the
+      // tray that causes the second system message with "clicked" set to false.
+      if (!notification.clicked || !notification.data) {
+        return Promise.resolve();
```

Comment:
I return ```Promise``` here just for the return result consistency and to make tests easier, but not sure if you like it :)